File: internal/memory/service.go
package memory

import (
	"context"
	"fmt"
	"time"

	"github.com/google/uuid"
)

// Embedder generates vector embeddings for text.
type Embedder interface {
	Embed(ctx context.Context, text string) ([]float64, error)
	EmbedBatch(ctx context.Context, texts []string) ([][]float64, error)
	Dimension() int
}

// Storage interface for memory persistence.
type Storage interface {
	Save(ctx context.Context, m *Memory) error
	Get(ctx context.Context, id string) (*Memory, error)
	List(ctx context.Context, opts ListOptions) ([]*Memory, error)
	Delete(ctx context.Context, id string) error
	Update(ctx context.Context, m *Memory) error
	SearchAllLayers(ctx context.Context, vector []float64, opts SearchOptions) ([]*SearchResult, error)
	TransferWorkingToEpisodic(ctx context.Context, sessionID string) (int, error)
	Close() error
}

// SearchOptions configures search behavior.
type SearchOptions struct {
	TopK            int
	MinScore        float64
	FilterLevels    []MemoryLevel
	IncludeObsolete bool
	SessionID       string
}

// SearchResult contains a memory and its similarity score.
type SearchResult struct {
	Memory *Memory
	Score  float64
}

// ListOptions configures list behavior.
type ListOptions struct {
	FilterLevels    []MemoryLevel
	IncludeObsolete bool
	Limit           int
	SortBy          string
	Reverse         bool
}

// CreateInput for creating new memories.
type CreateInput struct {
	Title     string
	Content   string
	Level     MemoryLevel
	Tags      []string
	SessionID string
	Source    string
	TaskID    string
	Author    string
}

// MemoryService provides memory operations.
type MemoryService struct {
	storage  Storage
	embedder Embedder
}

// NewMemoryService creates a new memory service.
func NewMemoryService(storage Storage, embedder Embedder) *MemoryService {
	return &MemoryService{storage: storage, embedder: embedder}
}

// Create creates a new memory.
func (s *MemoryService) Create(ctx context.Context, input CreateInput) (*Memory, error) {
	now := time.Now()
	m := &Memory{
		ID:      uuid.New().String(),
		Level:   input.Level,
		Title:   input.Title,
		Content: input.Content,
		Tags:    input.Tags,
		Context: MemoryContext{
			TaskID:    input.TaskID,
			SessionID: input.SessionID,
			Timestamp: now,
			Author:    input.Author,
			Tags:      input.Tags,
			Source:    input.Source,
		},
		CreatedAt: now,
		UpdatedAt: now,
		Obsolete:  false,
	}

	if err := m.Validate(); err != nil {
		return nil, err
	}

	text := fmt.Sprintf("Title: %s\n\nContent: %s", m.Title, m.Content)
	if len(m.Tags) > 0 {
		text += fmt.Sprintf("\n\nTags: %v", m.Tags)
	}
	embedding, err := s.embedder.Embed(ctx, text)
	if err != nil {
		return nil, fmt.Errorf("failed to generate embedding: %w", err)
	}
	m.Embedding = embedding

	if err := s.storage.Save(ctx, m); err != nil {
		return nil, fmt.Errorf("failed to save: %w", err)
	}
	return m, nil
}

// Search searches memories across all layers.
func (s *MemoryService) Search(ctx context.Context, query string, opts SearchOptions) ([]*SearchResult, error) {
	embedding, err := s.embedder.Embed(ctx, query)
	if err != nil {
		return nil, fmt.Errorf("failed to embed query: %w", err)
	}
	return s.storage.SearchAllLayers(ctx, embedding, opts)
}

// List lists memories with filtering.
func (s *MemoryService) List(ctx context.Context, opts ListOptions) ([]*Memory, error) {
	return s.storage.List(ctx, opts)
}

// Get retrieves a memory by ID.
func (s *MemoryService) Get(ctx context.Context, id string) (*Memory, error) {
	return s.storage.Get(ctx, id)
}

// Delete permanently deletes a memory.
func (s *MemoryService) Delete(ctx context.Context, id string) error {
	return s.storage.Delete(ctx, id)
}

// MarkObsolete soft-deletes a memory.
func (s *MemoryService) MarkObsolete(ctx context.Context, id string) error {
	m, err := s.storage.Get(ctx, id)
	if err != nil {
		return err
	}
	m.Obsolete = true
	m.UpdatedAt = time.Now()
	return s.storage.Update(ctx, m)
}

// TransferWorking transfers working memories to episodic.
func (s *MemoryService) TransferWorking(ctx context.Context, sessionID string) (int, error) {
	return s.storage.TransferWorkingToEpisodic(ctx, sessionID)
}

File: src/cortex/memory/service.py
"""Memory service — orchestrates storage and embeddings."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime
from uuid import uuid4

from cortex.embeddings.base import Embedder
from cortex.models.memory import Memory, MemoryContext, MemoryLevel, MemorySource
from cortex.models.results import SearchResult
from cortex.storage.base import ListOptions, SearchOptions, Storage


@dataclass
class CreateInput:
    """Input parameters for creating a memory."""

    title: str
    content: str
    level: MemoryLevel
    tags: list[str] = field(default_factory=list)
    session_id: str = ""
    source: MemorySource = MemorySource.manual
    task_id: str = ""
    author: str = ""


class MemoryService:
    """Orchestrates memory creation, retrieval, search and lifecycle.

    Responsibilities:
    - Generate embeddings for new/updated memories
    - Delegate persistence to the Storage backend
    - Manage LRU embedding cache on the Embedder
    """

    EMBED_TEXT_TEMPLATE = "Title: {title}\n\nContent: {content}\n\nTags: {tags}"

    def __init__(self, storage: Storage, embedder: Embedder) -> None:
        self._storage = storage
        self._embedder = embedder

    # ------------------------------------------------------------------
    # Write operations
    # ------------------------------------------------------------------

    def create(self, inp: CreateInput) -> Memory:
        """Create and persist a new memory with embedding."""
        context = MemoryContext(
            session_id=inp.session_id,
            task_id=inp.task_id,
            author=inp.author,
            source=inp.source,
            timestamp=datetime.now(UTC),
        )

        memory = Memory(
            id=str(uuid4()),
            level=inp.level,
            title=inp.title,
            content=inp.content,
            tags=inp.tags,
            context=context,
        )

        embed_text = self._embed_text(memory)
        memory.embedding = self._embedder.embed(embed_text)
        self._storage.save(memory)
        return memory

    def update(self, memory: Memory) -> Memory:
        """Update an existing memory (re-embeds if content changed)."""
        memory.touch()
        embed_text = self._embed_text(memory)
        memory.embedding = self._embedder.embed(embed_text)
        self._storage.update(memory)
        return memory

    def delete(self, memory_id: str) -> None:
        """Permanently delete a memory."""
        self._storage.delete(memory_id)

    def mark_obsolete(self, memory_id: str) -> Memory:
        """Soft-delete: mark a memory as obsolete."""
        memory = self._storage.get(memory_id)
        memory.obsolete = True
        memory.touch()
        self._storage.update(memory)
        return memory

    def transfer_working(self, session_id: str) -> int:
        """Transfer all working memories for a session to episodic layer."""
        return self._storage.transfer_working_to_episodic(session_id)

    def promote(self, memory_id: str, target_level: MemoryLevel) -> Memory:
        """Promote a memory to a higher layer (e.g. episodic → semantic)."""
        memory = self._storage.get(memory_id)
        memory.level = target_level
        memory.touch()
        self._storage.update(memory)
        return memory

    # ------------------------------------------------------------------
    # Read operations
    # ------------------------------------------------------------------

    def get(self, memory_id: str) -> Memory:
        """Retrieve a memory by ID or unique prefix."""
        return self._storage.get(memory_id)

    def list(self, opts: ListOptions | None = None) -> list[Memory]:
        """List memories with optional filtering."""
        return self._storage.list(opts)

    def search(
        self,
        query: str,
        opts: SearchOptions | None = None,
    ) -> list[SearchResult]:
        """Semantic search across all memory layers."""
        query_vector = self._embedder.embed(query)
        return self._storage.search_all_layers(query_vector, opts)

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    @classmethod
    def _embed_text(cls, memory: Memory) -> str:
        """Build the canonical embedding text for a memory."""
        tags_str = ", ".join(memory.tags) if memory.tags else ""
        return cls.EMBED_TEXT_TEMPLATE.format(
            title=memory.title,
            content=memory.content,
            tags=tags_str,
        )
